File: mage/src/main/java/mil/nga/giat/mage/map/TMSTileProvider.java
package mil.nga.giat.mage.map;

import android.util.Log;

import com.google.android.gms.maps.model.UrlTileProvider;

import java.net.MalformedURLException;
import java.net.URL;

import mil.nga.giat.mage.map.cache.URLCacheOverlay;

public class TMSTileProvider  extends UrlTileProvider {

    private static final String LOG_NAME = TMSTileProvider.class.getName();

    private final URLCacheOverlay myOverlay;

    public TMSTileProvider(int width, int height, URLCacheOverlay overlay) {
        super(width, height);

        myOverlay = overlay;
    }

    @Override
    public URL getTileUrl(int x, int y, int z) {
        String path = myOverlay.getURL().toString();
        path = path.replaceAll("\\{s\\}\\.", "");
        path = path.replaceAll("\\{x\\}", Integer.toString(x));
        long fixedY = (long) Math.pow(2, z) - y - 1;

        path = path.replaceAll("\\{y\\}", Long.toString(fixedY));
        path = path.replaceAll("\\{z\\}", Integer.toString(z));

        URL newPath = null;

        try {
            newPath = new URL(path);
        } catch (MalformedURLException e) {
            Log.w(LOG_NAME, "Problem with URL " + path, e);
        }

        return newPath;
    }
}

File: mage/src/main/java/mil/nga/giat/mage/map/TMSTileProvider.kt
package mil.nga.giat.mage.map

import android.util.Log
import com.google.android.gms.maps.model.UrlTileProvider
import mil.nga.giat.mage.map.cache.URLCacheOverlay
import java.net.MalformedURLException
import java.net.URL
import kotlin.math.pow

class TMSTileProvider(
   width: Int,
   height: Int,
   private val myOverlay: URLCacheOverlay
) : UrlTileProvider(width, height) {

   override fun getTileUrl(x: Int, y: Int, z: Int): URL? {
      val yTranslation = 2.0.pow(z.toDouble()).toLong() - y - 1

      val path = myOverlay.url.toString()
         .replace("{s}", "")
         .replace("{x}", x.toString())
         .replace("{y}", yTranslation.toString())
         .replace("{z}", z.toString())

      return try {
          URL(path)
      } catch (e: MalformedURLException) {
         Log.w(LOG_NAME, "Problem with URL $path", e)
         null
      }
   }

   companion object {
      private val LOG_NAME = TMSTileProvider::class.java.name
   }
}